[MRG+1] Bring back the green line

cbrnr opened the pull request:
Fixes #4771.

This (re)activates the vertical green line, which can be used to mark a specific position in time.

Current problems:
1. When right-clicking on a trace, both the green line is moved and the channel corresponding to the trace is marked as bad. Only the green line should be moved.
2. The green text indicating the current position of the line is too close to the x axis labels - is there any way to move the text more to the left?

@larsoner I think you are familiar with the plotting code, could you take a look? I'm not sure if what I did is OK or just an ugly hack - I'll ask specific question in code comments.

cbrnr commented on mne/viz/utils.py:
```diff
@@ -908,18 +908,22 @@ def _mouse_click(event, params):
-        _plot_annotations(raw, params)
-        params['plot_fun']()
-        return
+            if not isinstance(event.xdata, tuple):
```
It seems like `event.xdata` is sometimes a float and sometimes a tuple consisting of two floats. Any idea why?

larsoner replied:
Because at least two places in `master` we do:
```
_draw_vert_line(np.array([event.xdata] * 2), params)
```
i.e., convert it to a list with two elements. But again this should not be necessary once you use `vlines` instead of `plot` for the vertical line drawing.

cbrnr commented on mne/viz/utils.py:
```diff
@@ -985,9 +989,9 @@ def _find_channel_idx(ch_name, params):
-    params['ax_vertline'].set_data(xdata, np.array(params['ax'].get_ylim()))
-    params['ax_hscroll_vertline'].set_data(xdata, np.array([0., 1.]))
-    params['vertline_t'].set_text('%0.3f' % xdata[0])
+    params['ax_vertline'].set_data(xdata[0], np.array(params['ax'].get_ylim()))
```
Here `xdata` should now always be a tuple. It is strange that previously, in the first two lines of this function `xdata` was used, whereas in the third line `xdata[0]` was used. This can't be right.

larsoner replied:
... actually maybe it's sufficient only to update using a singleton. In any case I assume using `xdata[0]` and using `vlines` should be sufficient going forward

mmagnuski commented on mne/viz/raw.py:
```diff
@@ -445,6 +445,7 @@ def plot_raw(raw, events=None, duration=10.0, start=0.0, n_channels=20,
     params['callback_key'] = callback_key
     # have to store this, or it could get garbage-collected
     params['opt_button'] = opt_button
+    params['plot_vertline'] = partial(_draw_vert_line, params=params)
```
You don't seem to connect the _draw_vert_line function to 'button_press_event' but modify `_mouse_click`. Maybe it will be clearer if you attach another function to `'button_press_event'`?

mmagnuski replied:
ok, I now see that _mouse_click aggregates different mouse functionalities, so although I still think attaching multiple functions might be clearer it goes against how this code is composed.

cbrnr commented:
Thank you both for your comments. I have now opted to revert to the old behavior, which means left-clicking between the traces sets the green line (whereas left-clicking on a trace marks the channel as bad). Right-clicking does nothing, except in annotation mode where it deletes an annotation (so I guess we can use a right click in normal mode to show a topo in the future).

I have also moved the green text label farther to the left of the x axis labels (by appending two spaces to the string, I couldn't figure out a less hacky way).

Furthermore, it seems like the vertical line was updated in annotation mode on hover and when a segment was added/modified. I think this was probably not intended, so I have removed this. Let me know if I am wrong.

Otherwise, this looks good so far.

mmagnuski commented:
@cbrnr When one hovers over an annotation a line appears on annotation side (the side that the cursor is closer to), which informs the user about the interactivity of the annotations (you can drag their borders). Is this behavior still present?

cbrnr commented:
@mmagnuski yes, this still works. The hover function contained a reference to the green vertical line (and changed it), which I think was probably not what the person intended.

larsoner commented:
> (by appending two spaces to the string, I couldn't figure out a less hacky way).

Matplotlib usually has some way to customize the transformations, but two spaces sounds fine and only added two chars, compared to the multiple lines some transformation-based solution would likely require.

> (so I guess we can use a right click in normal mode to show a topo in the future).

In `evoked.plot` selecting a span of time with click-and-drag creates a topomap plot of the average activity in that window, so it would be good to replicate this behavior. But we don't need to do it here.

> it seems like the vertical line was updated in annotation mode on hover and when a segment was added/modified. I think this was probably not intended, so I have removed this. 

This does sound like it was unintended behavior. Did you test to make sure interactive annotation mode still works?

FYI I pushed a PEP8 fix / slightly cleaner way to iterate (using a short-circuiting generator expression rather than `np.any` on a list).

agramfort commented:
I get errors when playing with the GUI:

```
Traceback (most recent call last):
  File "/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 388, in process
    proxy(*args, **kwargs)
  File "/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 228, in __call__
    return mtd(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 950, in _mouse_click
    params['pick_bads_fun'](event)
  File "/Users/alex/work/src/mne-python/mne/viz/raw.py", line 99, in _pick_bad_channels
    params['info']['bads'] = _select_bads(event, params, bads)
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 999, in _select_bads
    _draw_vert_line(np.array([event.xdata] * 2), params)
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 991, in _draw_vert_line
    params['vertline_t'].set_text('%0.2f  ' % xdata)
TypeError: only size-1 arrays can be converted to Python scalars
```

I got this by putting a vertical line. Adding annotations, going back to normal mode and trying to put a vertical bar again. Can you replicate?

cbrnr commented:
@agramfort this was related to butterfly mode, I've fixed it. Could you please try again?

cbrnr commented:
@larsoner replicating existing behavior sounds like the way to go. @mmagnuski what do you think? Would it be OK to do it like `evoked.plot`?

Thanks for the cleanup, I did test going in and out of annotation mode and everything still works. There was a bug with the butterfly mode which I fixed. So this is ready to merge from my side.

massich commented:
Master:
![image](https://user-images.githubusercontent.com/7044835/47026029-daa54880-d164-11e8-9d42-8eb6530cc2b3.png)

PR:
![image](https://user-images.githubusercontent.com/7044835/47026359-8fd80080-d165-11e8-8818-362a52fb0913.png)

massich commented:
It was already working for me. But the PR makes it nicer. 

LGTM

agramfort commented:
good to go from my end
